modules/videoio/src/cap_msmf.cpp

```diff
@@ -997,71 +1236,181 @@ bool CvCapture_MSMF::openFinalize_(const VideoCaptureParameters* params)
+            return false;
+        } else
+        {
+            syncLastFrame = (value == 1) ? true : false;
+        }
+    }
+    return true;
+}
+
+bool CvCapture_MSMF::grabVideoFrame()
+{
+    DWORD streamIndex,  flags;
+    HRESULT hr;
+    videoSample.Release();
+    for (;;)
+    {
+        CV_TRACE_REGION("ReadSample");
+        if (!SUCCEEDED(hr = videoFileSource->ReadSample(
+            dwVideoStreamIndex, // Stream index.
+            0,             // Flags.
+            &streamIndex,  // Receives the actual stream index.
+            &flags,        // Receives status flags.
+            &sampleTime,   // Receives the time stamp.
+            &videoSample   // Receives the sample or NULL.
+        )))
+            break;
+        if (streamIndex != dwVideoStreamIndex)
+            break;
+        if (flags & (MF_SOURCE_READERF_ERROR | MF_SOURCE_READERF_ALLEFFECTSREMOVED | MF_SOURCE_READERF_ENDOFSTREAM))
+            break;
+        if (videoSample)
+            break;
+        if (flags & MF_SOURCE_READERF_STREAMTICK)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream tick detected. Retrying to grab the frame");
+        }
+    }
+    if (SUCCEEDED(hr))
+    {
+        if (streamIndex != dwVideoStreamIndex)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Wrong stream read. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ERROR)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream reading error. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ALLEFFECTSREMOVED)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream decoding error. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ENDOFSTREAM)
+        {
+            sampleTime += frameStep;
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): End of stream detected");
+        }
+        else
+        {
+            sampleTime += frameStep;
+            videoSample->GetSampleDuration(&curVideoTime);
+            allTime += curVideoTime;
+            if (duration - allTime < curVideoTime)
+            {
+                lastFrame = true;
+            }
+            if (flags & MF_SOURCE_READERF_NEWSTREAM)
+            {
+                CV_LOG_DEBUG(NULL, "videoio(MSMF): New stream detected");
+            }
+            if (flags & MF_SOURCE_READERF_NATIVEMEDIATYPECHANGED)
+            {
+                CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream native media type changed");
+            }
+            if (flags & MF_SOURCE_READERF_CURRENTMEDIATYPECHANGED)
+            {
+                CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream current media type changed");
+            }
+            return true;
+        }
+    }
+    return false;
+}
+
+bool CvCapture_MSMF::grabAudioFrame()
+{
+    DWORD streamIndex,  flags;
+    HRESULT hr;
+    std::vector<bool> outInstalFlag;
```

Comment:
Why do we need vector of flags here ? At the end they are just and-ed, mybe better use single boolean flag instead of vector ?